plugins/kotlin/src/main/java/org/quiltmc/quiltflower/kotlin/pass/ResugarKotlinMethodsPass.java

```diff
@@ -0,0 +1,139 @@
+        } else if (exprRes.expr != null) {
+          exprs.set(i, exprRes.expr);
+          res = true;
+        }
+      }
+    }
+
+    return res;
+  }
+
+  private static boolean resugarExprs(Exprent expr) {
+    boolean res = false;
+
+    for (Exprent ex : expr.getAllExprents()) {
+      res |= resugarExprs(ex);
+
+      Exprent map = resugarExpr(ex).expr;
+
+      if (map != null) {
+        expr.replaceExprent(ex, map);
+        res = true;
+      }
+    }
+
+    return res;
+  }
+
+
+  private static final MatchEngine[] NONNULL_INTRINSICS = {
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNullParameter(Ljava/lang/Object;Ljava/lang/String;)V"
+    ),
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNullExpressionValue(Ljava/lang/Object;Ljava/lang/String;)V"
+    ),
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNull(Ljava/lang/Object;Ljava/lang/String;)V"
+    ),
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNull(Ljava/lang/Object;)V"
+    )
+  };
+
+  // Intrinsics.areEqual($lhs$, $rhs$)
+  private static final MatchEngine EQUAL_INTRINSIC = new MatchEngine(
+    "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics name:areEqual parameter:0:$lhs$ parameter:1:$rhs$"
+  );
+
+  // ($x$ != null) ? $x$ : $y$
+  private static final MatchEngine TERNARY_NULL_CHECK = new MatchEngine(
+    "exprent type:function functype:ternary",
+    " exprent position:1 ret:$x1$",
+    " exprent position:2 ret:$y$",
+    " exprent position:0 type:function functype:neq",
+    "  exprent ret:$x$",
+    "  exprent type:constant consttype:null"
+  );
+
+  private static class ResugarRes {
+    public final Exprent expr;
+    public final boolean remove;
+
+    public ResugarRes(Exprent expr) {
+      this(expr, false);
+    }
+
+    public ResugarRes(Exprent expr, boolean remove) {
+      this.expr = expr;
+      this.remove = remove;
+    }
+  }
+
+  private static ResugarRes resugarExpr(Exprent ex) {
+    for (MatchEngine engine : NONNULL_INTRINSICS) {
+      if (engine.match(ex)) {
+        return new ResugarRes(null, true);
+      }
+    }
+
+    if (EQUAL_INTRINSIC.match(ex)) {
+      return new ResugarRes(new KFunctionExprent(FunctionExprent.FunctionType.EQ, List.of(
+        (Exprent) EQUAL_INTRINSIC.getVariableValue("$lhs$"), (Exprent) EQUAL_INTRINSIC.getVariableValue("$rhs$")
+      ), null), false);
+    }
+
+    if (TERNARY_NULL_CHECK.match(ex)) {
+      Exprent innerVal = (Exprent)TERNARY_NULL_CHECK.getVariableValue("$x1$");
+      if (innerVal instanceof InvocationExprent && ((InvocationExprent)innerVal).isUnboxingCall()) {
+        innerVal = ((InvocationExprent)innerVal).getInstance();
+      }
+
+      if (TERNARY_NULL_CHECK.getVariableValue("$x$").equals(innerVal)) {
```

Comment:
shouldn't there be a check that this is a variable?

Reply:
no I mean check if it's a var exprent?

Reply:
Oh right. good point